Role id backup restore

RichardJCai opened the pull request.

Fenil-P committed:
security, sql: Add User IDs in system tables and User Info struct

Previously we used usernames to access user information and privileges.
We will transition to using IDs by first adding user ids and updating the code
to use IDs where possible.

Release justification: Justified in RFC #77453
Release note: None

RichardJCai committed:
wip: fixing backup/restore for role id

RichardJCai committed:
fix backup/restore for users with role ids

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79559)
<!-- Reviewable:end -->

RichardJCai committed:
sql: GetUserID returns 0 if the role_id column does not exist

Release note: None

RichardJCai committed:
wip working on role_id query passing with conn exec

RichardJCai committed:
sql: Add role_id cache

Role ID cache maps SQLUsernames to RoleID.
Without the cache, we heavily hit the system.users table
for role ids which can be problematic.
Notably, TestWorkload under `schema_change_external_test.go` becomes
very slow and times out.

Release note: None
